src/main/java/dev/morling/onebrc/CalculateAverage_royvanrijn.java

```diff
@@ -112,210 +122,164 @@ private static long[] getSegments(int numberOfChunks) throws IOException {
-        // return nextAddress;
-    }
-
-    static final class Measurement {
-        int min, max, count;
-        long sum;
-
-        public Measurement() {
-            this.min = 1000;
-            this.max = -1000;
-        }
-
-        public Measurement updateWith(int measurement) {
-            min = min(min, measurement);
-            max = max(max, measurement);
+        public void updateWith(int measurement) {
+            min = Math.min(min, measurement);
+            max = Math.max(max, measurement);
             sum += measurement;
             count++;
-            return this;
         }
 
-        public Measurement updateWith(Measurement measurement) {
-            min = min(min, measurement.min);
-            max = max(max, measurement.max);
-            sum += measurement.sum;
-            count += measurement.count;
+        public Entry mergeWith(Entry entry) {
+            min = Math.min(min, entry.min);
+            max = Math.max(max, entry.max);
+            sum += entry.sum;
+            count += entry.count;
             return this;
         }
 
         public String toString() {
             return round(min) + "/" + round((1.0 * sum) / count) + "/" + round(max);
         }
 
-        private double round(double value) {
+        private static double round(double value) {
             return Math.round(value) / 10.0;
         }
     }
 
-    // branchless max (unprecise for large numbers, but good enough)
-    static int max(final int a, final int b) {
-        final int diff = a - b;
-        final int dsgn = diff >> 31;
-        return a - (diff & dsgn);
-    }
+    private static Entry createNewEntry(final long[] buffer, final long startAddress, final int lengthLongs, final int lengthBytes, final int temp) {
 
-    // branchless min (unprecise for large numbers, but good enough)
-    static int min(final int a, final int b) {
-        final int diff = a - b;
-        final int dsgn = diff >> 31;
-        return b + (diff & dsgn);
-    }
+        // --- This is a brand new entry, insert into the hashtable and do the extra calculations (once!) do slower calculations here.
+        final byte[] bytes = new byte[lengthBytes];
+        UNSAFE.copyMemory(null, startAddress, bytes, Unsafe.ARRAY_BYTE_BASE_OFFSET, lengthBytes);
+        final String city = new String(bytes, StandardCharsets.UTF_8);
 
-    private static int longHashStep(final int hash, final long word) {
-        return 31 * hash + (int) (word ^ (word >>> 32));
-    }
+        final long[] bufferCopy = new long[lengthLongs];
+        System.arraycopy(buffer, 0, bufferCopy, 0, lengthLongs);
 
-    private static long compilePattern(final byte value) {
-        return ((long) value << 56) | ((long) value << 48) | ((long) value << 40) | ((long) value << 32) |
-                ((long) value << 24) | ((long) value << 16) | ((long) value << 8) | (long) value;
+        // Add the entry:
+        return new Entry(bufferCopy, city, temp);
```

Comment:
I would try different things:
1. not create the String here, nor the byte[]
2. expose a method `String Entry::dataAsString(byte[] skratch)` which transform the data (it needs the "real" length, not the data.length) which use always the same byte[] provided from outside, to be used in the String creation
3. in the loop which iterate the entries (to merge them), allocate a byte[] upfront, and use `Entry::dataAsString` to create the keys

Given that we use EpsilonGC we try hard to avoid polluting with useless garbage the heap; during the runtime execution we already have all the required data to keep on processing, there's no need to allocate anything more than the Entry + its data.

Reply:
I can always construct the String from the long[], it contains the same bytes. I tried moving it back even further before, no idea why I undid that.

Reply:
Put it back again, only difference was a couple of 10ms, but then again, we're there right now that these matter.

Reply:
You measured JVM mode with EpsilonGC with these changes in? Because when we don't have a moving GC, allocating in the meantime can be relevant (or maybe not)